### ISSUE TITLE
Add beam search decoder with optional KenLM language model

### ISSUE THREAD

**Issue by PellelNitram** [PellelNitram/xournalpp_htr#120]
## Motivation

SimpleHTR currently uses greedy (best-path) CTC decoding. Beam search decoding — optionally with a language model — can improve word accuracy without retraining the model. The decoder should be parameterised so it can be treated as a hyperparameter during training sweeps, where each decoder tracks its own best checkpoint.

### Beam search decoder

Use [`pyctcdecode`](https://github.com/kensho-technologies/pyctcdecode) which supports plain beam search and beam search with a KenLM n-gram language model:

```python
from pyctcdecode import build_ctcdecoder

# Without LM
decoder = build_ctcdecoder(labels=charset)

# With LM
decoder = build_ctcdecoder(labels=charset, kenlm_model_path="lm.arff")

decoded = decoder.decode(log_probs_single)
```

### Language model

1. Gather a text corpus (e.g., IAM line transcriptions or general English text)
2. Train a KenLM n-gram model via the `kenlm` CLI → produces an `.arff` file
3. Pass it to `pyctcdecode`

### Code changes needed

1. **`network.py`** — add a `beam_decode()` function alongside `greedy_decode()`
2. **`train.py`** — `validate()` currently calls `greedy_decode()` to compute CER for early stopping / checkpoint selection. Run all configured decoders at each validation step, track separate best CER per decoder, save separate best checkpoints (e.g., `best_model_greedy.pth`, `best_model_beam.pth`)
3. **Inference paths** — `infer.py`, `demo.py` (import from `network.py`), and `inference_models.py` (inline greedy decode) need a decoder selection option
4. **Config** — add decoder config to `config.py` (decoder type, beam width, optional LM path)

### Decoder as hyperparameter

Since the decoder influences CER which drives early stopping and checkpoint selection, different decoders may pick different "best" epochs. Two options:

- **Parameterise**: add decoder to the sweep grid, one run per decoder config
- **Run all decoders per training**: evaluate with every decoder at each validation step, save a best checkpoint per decoder — trains once, gets optimal checkpoints for each decoder

**Comment by PellelNitram** [comment]
Branch: `120-add-beam-search-decoder-with-optional-kenlm-language-model`.

## Why this is being prioritised now

A benchmark sweep of all four pipelines (2026-09-18) shows **decoding, not detection, is the bottleneck**:

| Pipeline | Recall | CER (ci) | Word acc. |
|---|---|---|---|
| 2024-07-18_htr_pipeline | 69.7% | 39.4% | 37.4% |
| 2026-06-07_htr_pipeline_native | 71.6% | 42.9% | 35.1% |
| 2026-09-02_yolo_detector (in use) | 80.1% | 34.4% | 39.1% |
| 2026-09-17_rf_detr_detector | 66.8% | 36.4% | 41.8% |

A 13.3 pp swing in detector recall (66.8 → 80.1) barely moves word accuracy — every pipeline sits in a 35–42% band because CER is pinned at 34–43%. Improving the decoder attacks the term that actually dominates.

**Baseline to beat** (`2026-09-02_yolo_detector`): CER_ci **34.4%**, word accuracy **39.1%**, recall × (1 − CER_ci) **52.5%**.

## The cheap path to a measurement

`SimpleHTRModel.recognize()` (`xournalpp_htr/inference_models.py:165`) already runs the ONNX graph and gets `log_probs` of shape `(seq_len, batch, num_classes)`, then does an inline greedy collapse. **`pyctcdecode` can be dropped straight in there — no retraining and no re-export.** So the end-to-end benchmark impact can be measured before touching `train.py` or the checkpoint-selection work described above:

```bash
uv run python scripts/run_benchmark.py -p 2026-09-02_yolo_detector
```

Suggested order: prove the win at inference first, then decide whether the per-decoder checkpoint-selection machinery in `train.py` is worth building.

**Watch the blank index.** This implementation puts the CTC blank at `len(charset)` (i.e. last) — see the `blank = len(self._charset)` line. `pyctcdecode`'s label/blank convention differs and must be reconciled, or output will be silently wrong rather than erroring.

## Verified code locations

- `xournalpp_htr/training/simple_htr/network.py:104` — `greedy_decode()`, where `beam_decode()` belongs
- `xournalpp_htr/training/simple_htr/train.py:159` — `validate()` calls `greedy_decode()` for CER → early stopping / checkpoint selection
- `xournalpp_htr/training/simple_htr/infer.py:50` — imports and calls `greedy_decode()`
- `xournalpp_htr/inference_models.py:165` — `SimpleHTRModel.recognize()`, inline greedy decode (no import from `network.py`)
- `xournalpp_htr/training/simple_htr/config.py` — Hydra config, where decoder settings go
- `pyctcdecode`/`kenlm` belong in the `training-simple-htr` extra in `pyproject.toml`; anything needed at *inference* must go in the base deps, since inference stays lean per ADR 006

## Environment

- Local MacBook cannot run the test suite (`htr_pipeline/models/reader.onnx` missing from the venv) and has no outbound network for `uv`. Run benchmarks and training on the GPU VM.
- The benchmark is fast (~10–25 s per pipeline), so iterating on decoder settings against it is cheap.
